### ISSUE TITLE
STY/DEV: fix and enable lint rule UP038

### ISSUE THREAD

**Issue by lucascolley** [scipy/scipy#21547]
gh-21545 disabled lint rule UP038, which enforces use of e.g. `isinstance(x, int | float | bool)` rather than `isinstance(x, (int, float, bool))`. I agree with the rule developers that the former is less ambiguous, as the latter is reasonably interpreted as "a tuple of an int, a float, and a bool" rather than "an int, or a float, or a bool". Therefore, it would be good to fix this rule throughout the codebase and re-enable it in the linter.

Task:
- remove `"UP038"` from the ignore list in https://github.com/scipy/scipy/blob/34a2b3687c1587c094c5f0d3314dd06d1c100cdc/tools/lint.toml#L23
- fix up all errors from ruff

Here is the output of `ruff check . --target-version=py310 --select UP038` at the time of writing:

<details> <summary> dropdown </summary>

```
(scipy:lint) lucascolley@arm64-apple-darwin20 scipy % ruff check . --target-version=py310 --select UP038
doc/source/conf.py:461:8: UP038 Use `X | Y` in `isinstance` call instead of `(X, Y)`
    |
459 |     # SciPy's distributions are instances of *_gen. Point to this
460 |     # class since it contains the implementation of all the methods.
461 |     if isinstance(obj, (rv_generic, multi_rv_generic)):
    |        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ UP038
462 |         obj = obj.__class__
463 |     try:
    |
    = help: Convert to `X | Y`

scipy/_build_utils/system_info.py:26:20: UP038 Use `X | Y` in `isinstance` call instead of `(X, Y)`
   |
24 |             if new_dict.get(key, None) is not None:
25 |                 old_value = new_dict[key]
26 |                 if isinstance(value, (list, tuple)):
   |                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ UP038
27 |                     if isinstance(old_value, (list, tuple)):
28 |                         new_dict[key] = list(old_value) + list(value)
   |
   = help: Convert to `X | Y`

scipy/_build_utils/system_info.py:27:24: UP038 Use `X | Y` in `isinstance` call instead of `(X, Y)`
   |
25 |                 old_value = new_dict[key]
26 |                 if isinstance(value, (list, tuple)):
27 |                     if isinstance(old_value, (list, tuple)):
   |                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ UP038
28 |                         new_dict[key] = list(old_value) + list(value)
29 |                         continue
   |
   = help: Convert to `X | Y`

scipy/_lib/_array_api.py:81:12: UP038 Use `X | Y` in `isinstance` call instead of `(X, Y)`
   |
79 |         elif isinstance(array, np.matrix):
80 |             raise TypeError("Inputs of type `numpy.matrix` are not supported.")
81 |         if isinstance(array, (np.ndarray, np.generic)):
   |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ UP038
82 |             dtype = array.dtype
83 |             if not (np.issubdtype(dtype, np.number) or np.issubdtype(dtype, np.bool_)):
   |
   = help: Convert to `X | Y`

scipy/_lib/_docscrape.py:706:24: UP038 Use `X | Y` in `isinstance` call instead of `(X, Y)`
    |
704 |                 and (
705 |                     func is None
706 |                     or isinstance(func, (property, cached_property))
    |                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ UP038
707 |                     or inspect.isdatadescriptor(func)
708 |                 )
    |
    = help: Convert to `X | Y`

scipy/_lib/array_api_compat/array_api_compat/common/_aliases.py:281:16: UP038 Use `X | Y` in `isinstance` call instead of `(X, Y)`
    |
279 | ) -> ndarray:
280 |     def _isscalar(a):
281 |         return isinstance(a, (int, float, type(None)))
    |                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ UP038
282 |     min_shape = () if _isscalar(min) else min.shape
283 |     max_shape = () if _isscalar(max) else max.shape
    |
    = help: Convert to `X | Y`

scipy/_lib/array_api_compat/array_api_compat/common/_helpers.py:63:13: UP038 Use `X | Y` in `isinstance` call instead of `(X, Y)`
   |
62 |     # TODO: Should we reject ndarray subclasses?
63 |     return (isinstance(x, (np.ndarray, np.generic))
   |             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ UP038
64 |             and not _is_jax_zero_gradient_array(x))
   |
   = help: Convert to `X | Y`

scipy/_lib/array_api_compat/array_api_compat/common/_helpers.py:94:12: UP038 Use `X | Y` in `isinstance` call instead of `(X, Y)`
   |
93 |     # TODO: Should we reject ndarray subclasses?
94 |     return isinstance(x, (cp.ndarray, cp.generic))
   |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ UP038
95 | 
96 | def is_torch_array(x):
   |
   = help: Convert to `X | Y`

scipy/cluster/hierarchy.py:2980:14: UP038 Use `X | Y` in `isinstance` call instead of `(X, Y)`
     |
2978 |         # reset to its default
2979 |         palette = _link_line_colors_default
2980 |     elif not isinstance(palette, (list, tuple)):
     |              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ UP038
2981 |         raise TypeError("palette must be a list or tuple")
2982 |     _ptypes = [isinstance(p, str) for p in palette]
     |
     = help: Convert to `X | Y`

scipy/cluster/hierarchy.py:3284:8: UP038 Use `X | Y` in `isinstance` call instead of `(X, Y)`
     |
3282 |     Zs = Z.shape
3283 |     n = Zs[0] + 1
3284 |     if isinstance(p, (int, float)):
     |        ^^^^^^^^^^^^^^^^^^^^^^^^^^^ UP038
3285 |         p = int(p)
3286 |     else:
     |
     = help: Convert to `X | Y`

scipy/datasets/_utils.py:32:16: UP038 Use `X | Y` in `isinstance` call instead of `(X, Y)`
   |
30 |         shutil.rmtree(cache_dir)
31 |     else:
32 |         if not isinstance(datasets, (list, tuple)):
   |                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ UP038
33 |             # single dataset method passed should be converted to list
34 |             datasets = [datasets, ]
   |
   = help: Convert to `X | Y`

scipy/integrate/_quad_vec.py:350:16: UP038 Use `X | Y` in `isinstance` call instead of `(X, Y)`
    |
349 |         if global_integral is None:
350 |             if isinstance(ig, (float, complex)):
    |                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ UP038
351 |                 # Specialize for scalars
352 |                 if norm_func in (_max_norm, np.linalg.norm):
    |
    = help: Convert to `X | Y`

scipy/interpolate/_fitpack2.py:2285:14: UP038 Use `X | Y` in `isinstance` call instead of `(X, Y)`
     |
2283 |         if pole_values is None:
2284 |             pole_values = (None, None)
2285 |         elif isinstance(pole_values, (float, np.float32, np.float64)):
     |              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ UP038
2286 |             pole_values = (pole_values, pole_values)
2287 |         if isinstance(pole_continuity, bool):
     |
     = help: Convert to `X | Y`

scipy/interpolate/_interpolate.py:1653:16: UP038 Use `X | Y` in `isinstance` call instead of `(X, Y)`
     |
1651 |             orders = [None] * m
1652 |         else:
1653 |             if isinstance(orders, (int, np.integer)):
     |                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ UP038
1654 |                 orders = [orders] * m
1655 |             k = max(k, max(orders))
     |
     = help: Convert to `X | Y`

scipy/io/_netcdf.py:566:16: UP038 Use `X | Y` in `isinstance` call instead of `(X, Y)`
    |
565 |             # bytes index into scalars in py3k. Check for "string" types
566 |             if isinstance(values, (str, bytes)):
    |                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ UP038
567 |                 sample = values
568 |             else:
    |
    = help: Convert to `X | Y`

scipy/ndimage/_ni_support.py:83:10: UP038 Use `X | Y` in `isinstance` call instead of `(X, Y)`
   |
81 |             complex_type = np.promote_types(input.dtype, np.complex64)
82 |             output = np.zeros(shape, dtype=complex_type)
83 |     elif isinstance(output, (type, np.dtype)):
   |          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ UP038
84 |         # Classes (like `np.float32`) and dtypes are interpreted as dtype
85 |         if complex_output and np.dtype(output).kind != 'c':
   |
   = help: Convert to `X | Y`

scipy/ndimage/_support_alternative_backends.py:39:20: UP038 Use `X | Y` in `isinstance` call instead of `(X, Y)`
   |
37 |                 result = func(*args, **kwds)
38 | 
39 |                 if isinstance(result, (np.ndarray, np.generic)):
   |                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ UP038
40 |                     # XXX: np.int32->np.array_0D
41 |                     return xp.asarray(result)
   |
   = help: Convert to `X | Y`

scipy/optimize/_minpack_py.py:934:8: UP038 Use `X | Y` in `isinstance` call instead of `(X, Y)`
    |
932 |         ydata = np.asarray(ydata, float)
933 | 
934 |     if isinstance(xdata, (list, tuple, np.ndarray)):
    |        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ UP038
935 |         # `xdata` is passed straight to the user-defined `f`, so allow
936 |         # non-array_like `xdata`.
    |
    = help: Convert to `X | Y`

scipy/optimize/_root_scalar.py:274:16: UP038 Use `X | Y` in `isinstance` call instead of `(X, Y)`
    |
273 |     if meth in ['bisect', 'ridder', 'brentq', 'brenth', 'toms748']:
274 |         if not isinstance(bracket, (list, tuple, np.ndarray)):
    |                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ UP038
275 |             raise ValueError(f'Bracket needed for {method}')
    |
    = help: Convert to `X | Y`

scipy/optimize/_shgo_lib/_complex.py:140:20: UP038 Use `X | Y` in `isinstance` call instead of `(X, Y)`
    |
138 |             self.g_cons = []
139 |             self.g_args = []
140 |             if not isinstance(constraints, (tuple, list)):
    |                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ UP038
141 |                 constraints = (constraints,)
    |
    = help: Convert to `X | Y`

scipy/optimize/tests/test__shgo.py:29:16: UP038 Use `X | Y` in `isinstance` call instead of `(X, Y)`
   |
27 |     cons = []
28 |     if g is not None:
29 |         if not isinstance(g, (tuple, list)):
   |                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^ UP038
30 |             g = (g,)
31 |         else:
   |
   = help: Convert to `X | Y`

scipy/signal/_ltisys.py:1357:16: UP038 Use `X | Y` in `isinstance` call instead of `(X, Y)`
     |
1356 |       def _check_binop_other(self, other):
1357 |           return isinstance(other, (StateSpace, np.ndarray, float, complex,
     |  ________________^
1358 | |                                   np.number, int))
     | |__________________________________________________^ UP038
1359 |   
1360 |       def __mul__(self, other):
     |
     = help: Convert to `X | Y`

scipy/signal/_ltisys.py:1898:18: UP038 Use `X | Y` in `isinstance` call instead of `(X, Y)`
     |
1897 |     no_input = (U is None or
1898 |                 (isinstance(U, (int, float)) and U == 0.) or
     |                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^ UP038
1899 |                 not np.any(U))
     |
     = help: Convert to `X | Y`

scipy/signal/_ltisys.py:2263:12: UP038 Use `X | Y` in `isinstance` call instead of `(X, Y)`
     |
2261 |     """
2262 |     if isinstance(system, lti):
2263 |         if isinstance(system, (TransferFunction, ZerosPolesGain)):
     |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ UP038
2264 |             sys = system
2265 |         else:
     |
     = help: Convert to `X | Y`

scipy/signal/_ltisys.py:3398:12: UP038 Use `X | Y` in `isinstance` call instead of `(X, Y)`
     |
3396 |         system = system._as_tf()
3397 | 
3398 |     if not isinstance(system, (TransferFunction, ZerosPolesGain)):
     |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ UP038
3399 |         raise ValueError('Unknown system type')
     |
     = help: Convert to `X | Y`

scipy/sparse/_sputils.py:262:14: UP038 Use `X | Y` in `isinstance` call instead of `(X, Y)`
    |
261 | def issequence(t) -> bool:
262 |     return ((isinstance(t, (list, tuple)) and
    |              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^ UP038
263 |             (len(t) == 0 or np.isscalar(t[0]))) or
264 |             (isinstance(t, np.ndarray) and (t.ndim == 1)))
    |
    = help: Convert to `X | Y`

scipy/sparse/_sputils.py:268:14: UP038 Use `X | Y` in `isinstance` call instead of `(X, Y)`
    |
267 | def ismatrix(t) -> bool:
268 |     return ((isinstance(t, (list, tuple)) and
    |              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^ UP038
269 |              len(t) > 0 and issequence(t[0])) or
270 |             (isinstance(t, np.ndarray) and t.ndim == 2))
    |
    = help: Convert to `X | Y`

scipy/sparse/linalg/_matfuncs.py:599:8: UP038 Use `X | Y` in `isinstance` call instead of `(X, Y)`
    |
598 |     # Avoid indiscriminate asarray() to allow sparse or other strange arrays.
599 |     if isinstance(A, (list, tuple, np.matrix)):
    |        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ UP038
600 |         A = np.asarray(A)
601 |     if len(A.shape) != 2 or A.shape[0] != A.shape[1]:
    |
    = help: Convert to `X | Y`

scipy/sparse/tests/test_base.py:71:16: UP038 Use `X | Y` in `isinstance` call instead of `(X, Y)`
   |
69 |         arrays = []
70 |         for a in x.__dict__.values():
71 |             if isinstance(a, (np.ndarray, np.generic)):
   |                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ UP038
72 |                 arrays.append(a)
73 |         return arrays
   |
   = help: Convert to `X | Y`

scipy/sparse/tests/test_base.py:346:16: UP038 Use `X | Y` in `isinstance` call instead of `(X, Y)`
    |
344 |                                      (datsp == np.nan).toarray())
345 | 
346 |         if not isinstance(self, (TestBSR, TestCSC, TestCSR)):
    |                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ UP038
347 |             pytest.skip("Bool comparisons only implemented for BSR, CSC, and CSR.")
348 |         for dtype in self.checked_dtypes:
    |
    = help: Convert to `X | Y`

scipy/sparse/tests/test_base.py:385:16: UP038 Use `X | Y` in `isinstance` call instead of `(X, Y)`
    |
383 |                                      (datsp != np.nan).toarray())
384 | 
385 |         if not isinstance(self, (TestBSR, TestCSC, TestCSR)):
    |                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ UP038
386 |             pytest.skip("Bool comparisons only implemented for BSR, CSC, and CSR.")
387 |         for dtype in self.checked_dtypes:
    |
    = help: Convert to `X | Y`

scipy/sparse/tests/test_base.py:448:16: UP038 Use `X | Y` in `isinstance` call instead of `(X, Y)`
    |
446 |             assert_array_equal_dtype(dat < datsp2, datsp < dat2)
447 | 
448 |         if not isinstance(self, (TestBSR, TestCSC, TestCSR)):
    |                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ UP038
449 |             pytest.skip("Bool comparisons only implemented for BSR, CSC, and CSR.")
450 |         for dtype in self.checked_dtypes:
    |
    = help: Convert to `X | Y`

scipy/sparse/tests/test_base.py:510:16: UP038 Use `X | Y` in `isinstance` call instead of `(X, Y)`
    |
508 |             assert_array_equal_dtype(dat > datsp2, datsp > dat2)
509 | 
510 |         if not isinstance(self, (TestBSR, TestCSC, TestCSR)):
    |                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ UP038
511 |             pytest.skip("Bool comparisons only implemented for BSR, CSC, and CSR.")
512 |         for dtype in self.checked_dtypes:
    |
    = help: Convert to `X | Y`

scipy/sparse/tests/test_base.py:568:16: UP038 Use `X | Y` in `isinstance` call instead of `(X, Y)`
    |
566 |             assert_array_equal_dtype(dat <= datsp2, datsp <= dat2)
567 | 
568 |         if not isinstance(self, (TestBSR, TestCSC, TestCSR)):
    |                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ UP038
569 |             pytest.skip("Bool comparisons only implemented for BSR, CSC, and CSR.")
570 |         for dtype in self.checked_dtypes:
    |
    = help: Convert to `X | Y`

scipy/sparse/tests/test_base.py:626:16: UP038 Use `X | Y` in `isinstance` call instead of `(X, Y)`
    |
624 |             assert_array_equal_dtype(dat >= datsp2, datsp >= dat2)
625 | 
626 |         if not isinstance(self, (TestBSR, TestCSC, TestCSR)):
    |                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ UP038
627 |             pytest.skip("Bool comparisons only implemented for BSR, CSC, and CSR.")
628 |         for dtype in self.checked_dtypes:
    |
    = help: Convert to `X | Y`

scipy/sparse/tests/test_base.py:5204:12: UP038 Use `X | Y` in `isinstance` call instead of `(X, Y)`
     |
5202 |     def _compare_index_dtype(self, m, dtype):
5203 |         dtype = np.dtype(dtype)
5204 |         if isinstance(m, (csc_matrix, csr_matrix, bsr_matrix)):
     |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ UP038
5205 |             return (m.indices.dtype == dtype) and (m.indptr.dtype == dtype)
5206 |         elif isinstance(m, coo_matrix):
     |
     = help: Convert to `X | Y`

scipy/stats/_distn_infrastructure.py:640:12: UP038 Use `X | Y` in `isinstance` call instead of `(X, Y)`
    |
638 |     # np.atleast_1d returns an array if only one argument, or a list of arrays
639 |     # if more than one argument.
640 |     if not isinstance(newargs, (list, tuple)):
    |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ UP038
641 |         newargs = (newargs,)
    |
    = help: Convert to `X | Y`

scipy/stats/_qmc.py:80:24: UP038 Use `X | Y` in `isinstance` call instead of `(X, Y)`
   |
79 |     """
80 |     if seed is None or isinstance(seed, (numbers.Integral, np.integer)):
   |                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ UP038
81 |         return np.random.default_rng(seed)
82 |     elif isinstance(seed, (np.random.RandomState, np.random.Generator)):
   |
   = help: Convert to `X | Y`

scipy/stats/_qmc.py:82:10: UP038 Use `X | Y` in `isinstance` call instead of `(X, Y)`
   |
80 |     if seed is None or isinstance(seed, (numbers.Integral, np.integer)):
81 |         return np.random.default_rng(seed)
82 |     elif isinstance(seed, (np.random.RandomState, np.random.Generator)):
   |          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ UP038
83 |         return seed
84 |     else:
   |
   = help: Convert to `X | Y`

scipy/stats/_sampling.py:1031:16: UP038 Use `X | Y` in `isinstance` call instead of `(X, Y)`
     |
1030 |         """
1031 |         if not isinstance(size, (numbers.Integral, np.integer)):
     |                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ UP038
1032 |             raise ValueError("size must be an integer.")
1033 |         # urng will be used to draw the samples for testing the error
     |
     = help: Convert to `X | Y`

scipy/stats/_stats_py.py:444:12: UP038 Use `X | Y` in `isinstance` call instead of `(X, Y)`
    |
443 |     """
444 |     if not isinstance(p, (int, float)):
    |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^ UP038
445 |         raise ValueError("Power mean only defined for exponent of type int or "
446 |                          "float.")
    |
    = help: Convert to `X | Y`

Found 41 errors.
```

</details>

---

By the way, `ruff check . --config=tools/lint.toml --exclude tools/ --exclude scipy/_lib/unuran --exclude scipy/_lib/highs --exclude scipy/_lib/array_api_compat --exclude scipy/_lib/cobyqa --exclude scipy/_lib/_uarray --exclude scipy/_lib/_docscrape.py --exclude scipy/_build_utils/tempita` only returns 54 errors, so a follow-up PR could deal with the remaining ~13 once UP038 is done.

**Comment by fancidev** [comment]
I think the proposed syntax is nice as it is consistent with the recommended way of type annotation (i.e. `X | Y` rather than `typing.Union[X, Y]`).

However, the [doc](https://docs.python.org/3/library/functions.html#isinstance) of the `isinstance` function mentions that union type (the syntax suggested by this PR) is only supported since Python 3.10. And [this page](https://devguide.python.org/versions/) says Python 3.9 will remain supported until October, 2025. Therefore, if SciPy is supposed to work with the minimum supported Python version, I guess the code base cannot be changed until after Python 3.9’s lifecycle.

**Comment by tylerjereddy** [comment]
> Therefore, if SciPy is supposed to work with the minimum supported Python version, I guess the code base cannot be changed until after Python 3.9’s lifecycle.

we're good on this part at least--we've already dropped Python `3.9.x` support in `main` and even the recent stable release series (`1.14.x` is `>=3.10.x`).

There are a bunch of standards documents we follow to some degree like:

https://numpy.org/neps/nep-0029-deprecation_policy.html
https://scientific-python.org/specs/spec-0000/

More annoying is that I thought @WarrenWeckesser had noted the new syntax is actually *slower* to execute.

**Comment by mdhaber** [comment]
> the new syntax is actually slower to execute.

Yes, that was also discussed at the NumFOCUS summit.
```python3
x = 1

%timeit isinstance(x, (float, complex))
# 209 ns ± 4.09 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)

%timeit isinstance(x, float | complex)
# 474 ns ± 133 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)
```

But of course, the difference is pretty small compared to many NumPy function calls.

```python3
import numpy as np
x = np.asarray(x)
%timeit np.issubdtype(x.dtype, np.inexact)
# 1.14 µs ± 287 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)
```

**Comment by j-bowhay** [comment]
I have to admit I also share reservations about this, is it really an improvement?
